### PR TITLE
fix #13393 better hash for primitive types, avoiding catastrophic (1000x) slowdowns for certain input distributions

### LINKED ISSUE #13393: Slow insertion of uint in Tables and Sets Collections depending on the value
A long time ago I noticed that the Tables and Sets collections took a long time to add values ​​and I decided to search further.

So, I noticed that the problem was: the higher a significant bit in an integer, the slower the insertion of a value will be.

Using uint32, entering values ​​with the lowest 16 bits is 14 times faster than entering values ​​with the highest 16 bits.

Using uint64 the situation is more serious. Entering values ​​with the low 32 bits is about 256 times faster than entering values ​​with the highest 32 bits, this for the Sets collection. For Tables collection, the difference increases to 348 times.

I will not put sample code, because I created a github repository with the problem, where it is possible to see the codes used and even a quick solution using seqs and sort, as well as code using similar collections in other languages ​​to do the same work that show a greater consistency in execution time.

It is a problem related or equal to the one treated here: [11764](https://github.com/nim-lang/Nim/issues/11764)

### Example and Current Output
Visit this link [https://github.com/rockcavera/nim-problem](https://github.com/rockcavera/nim-problem)

### Expected Output
A less significant difference is expected between the addition of values, that is, closer execution times, the same occurs in other languages.

### Possible Solution

* I can't say if the problem is with the designer adopted for the collections or a wrong choice of hash... I didn't delve into the current code of the collections due to lack of time.

### Additional Information

* The test was carried out on several Nim versions since before the release of 1.0 and they all have the same problem

### PR DESCRIPTION
**Pull request by timotheecour**
* revives https://github.com/nim-lang/Nim/pull/11767, same idea but further improved
* fixes https://github.com/nim-lang/Nim/issues/13393

using the tests provided by @rockcavera in https://github.com/rockcavera/nim-problem:

```
nim c -r -d:danger nim-problem/codesetsu64.nim
```
now completes in 0.075 seconds instead of 75.4 seconds prior to this PR, **a speedup of 1000X** for this example, which is exactly consistent with what I had already observed in https://github.com/nim-lang/Nim/pull/11767 on my own benchmark, noting a 100 to 1000 speedup (https://github.com/timotheecour/vitanim/blob/master/testcases/tests/t0129b.nim)

## before PR
* before https://github.com/nim-lang/Nim/pull/12407, the hash for primitive types was `hash(x)=x bitand 2^n-1` which is a terrible hash resulting in lots of (trivial) collisions, ignoring all high order bits.
* after https://github.com/nim-lang/Nim/pull/12407 (which aimed at addressing https://github.com/nim-lang/Nim/issues/11764) the hash was `cast[Hash](cast[uint](x) * 11)` which is still not a good hash, resulting in lots of collisions depending on input distributions, as demonstrated by https://github.com/nim-lang/Nim/issues/13393; for example when input distributions only vary in high order bits, the hash function collides to a single point
* floats were problematic because of formula used (1.0 + x), causing all values to collide if x was small enough (x+1.0 would collapse to 1.0 for x small enough due to FP semantics)

## after PR
* all primitive types (including pointer etc) with sizeof(x) in {4,8} pass through either a hashUInt64 or hashUInt32 specialized hash that has good hashing properties (cascading effect), as commonly prescribed in articles discussing hashing functions
* sizeof(x)<4 was faster using the preexisting identity hash so the fix checks for that sizeof(x)>=4 criterion; indeed in this case the number of unique values is limited so collisions are less of an issue.
* float is now properly handled; in particular denormalization is checked so that hash(0.0)==hash(-0.0), and also removes the previous hack based on `x+1.0` which caused collisions
* performance regression test is added, testing various input distributions, and for nim c, cpp, js, as well as VM for those modes.

* It's a tricky problem to get it right: it has to work on 32 bit, 64 bit, with js, c, cpp, with VM during defined(js) (where sizeof(int)=4 even on 64bit), etc; and it has be be performant in all cases.

### CODE REVIEW

File: lib/pure/hashes.nim [hunk 1]
```diff
@@ -103,7 +144,14 @@ proc hash*(x: pointer): Hash {.inline.} =
       }
     """
   else:
-    result = cast[Hash](cast[uint](x) shr 3) # skip the alignment
+    #[
+    3 bug fixes:
+    * s/cast[Hash]()/hash()/ (#11764)
+    * s/uint/BiggestInt/
+    * s/x shr 3/x/ (no reason to skip the alignment)
+    # note that we can't use unsigned because nimscript doesn't have `$`(uint)
```
Comment (Clyybber) [inline lib/pure/hashes.nim]: Comments are not commit messages.

File: lib/pure/hashes.nim [hunk 2]
```diff
@@ -103,7 +144,14 @@ proc hash*(x: pointer): Hash {.inline.} =
       }
     """
   else:
-    result = cast[Hash](cast[uint](x) shr 3) # skip the alignment
+    #[
+    3 bug fixes:
+    * s/cast[Hash]()/hash()/ (#11764)
+    * s/uint/BiggestInt/
+    * s/x shr 3/x/ (no reason to skip the alignment)
```
Comment (Clyybber) [inline lib/pure/hashes.nim]: Comments are not commit messages

**Comment by Clyybber** [comment]
@timotheecour Does this also work for the JS backend?

**Review by timotheecour — COMMENTED** [review]
(can't delete this comment, was a duplicate of https://github.com/nim-lang/Nim/pull/13410#discussion_r378838293)

File: lib/pure/hashes.nim (hunk 2 above)
Comment (timotheecour) [reply lib/pure/hashes.nim]: sure but I wanted to show rationale for the new formula; i've now removed these comments for the most part

**Comment by timotheecour** [comment]
> @timotheecour Does this also work for the JS backend?

yes the js backend also improves hashes for primitive types, with improved randomness:

```nim
import hashes, strformat

template hash2(x) =
  block:
    let s {.inject.} = astToStr(x)
    let j {.inject.} = hash(x)
    echo &"""{s:>8} => {j}"""

hash2(0.3)
hash2("foo")
hash2(123)
hash2(124)
hash2(125)
```

```nim
import os,strformat
proc main() =
  let file = currentSourcePath / "../t10210.nim"
  for nim in @["/Users/timothee/git_clone//nim//Nim_devel//bin/nim", "/Users/timothee/git_clone//nim//Nim_prs//bin/nim.pr_fix_11764_hash_revived"]:
    for mode in @["c", "js"]:
      let cmd = fmt"{nim} {mode} -r --hints:off {file}"
      echo cmd
      doAssert execShellCmd(cmd) == 0
main()
```

prints:
```
/Users/timothee/git_clone//nim//Nim_devel//bin/nim c -r --hints:off /Users/timothee/git_clone/nim/timn/tests/nim/all/t10210.nim
     0.3 => 4608533498688228557
   "foo" => 4138058784
     123 => 1353
     124 => 1364
     125 => 1375
/Users/timothee/git_clone//nim//Nim_devel//bin/nim js -r --hints:off /Users/timothee/git_clone/nim/timn/tests/nim/all/t10210.nim
     0.3 => 1.3
   "foo" => -156908512
     123 => 1353
     124 => 1364
     125 => 1375
/Users/timothee/git_clone//nim//Nim_prs//bin/nim.pr_fix_11764_hash_revived c -r --hints:off /Users/timothee/git_clone/nim/timn/tests/nim/all/t10210.nim
     0.3 => -7272369355435918195
   "foo" => 4138058784
     123 => -508526132817443961
     124 => -1530243182942418667
     125 => -6335688243927128111
/Users/timothee/git_clone//nim//Nim_prs//bin/nim.pr_fix_11764_hash_revived js -r --hints:off /Users/timothee/git_clone/nim/timn/tests/nim/all/t10210.nim
     0.3 => 743435112
   "foo" => -156908512
     123 => -1632341525
     124 => -773518108
     125 => -163873810
```

note that hashes for `nim js` are not guaranteed to be same as hashes for `nim c`, but hashes for vm (whether running in `nim js` or `nim c`) are same as hashes for `nim c`; this behavior is pre-existing to this PR and is really hard/impossible to fix until https://github.com/nim-lang/RFCs/issues/187 is addressed (if anything, because js uses a float type for numbers and can't represent int64)

File: lib/pure/hashes.nim (hunk 2 above)
Comment (Clyybber) [reply lib/pure/hashes.nim]: I understand. But rationale can be explained in the commit messages or PR description : )
Better than get the code all messy, since these are not really explanations but seds

File: lib/pure/hashes.nim [hunk 3]
```diff
@@ -86,6 +89,50 @@ proc hashData*(data: pointer, size: int): Hash =
...
+  else:
+    when defined(js):
+      # could use BigInt, see https://github.com/nim-lang/RFCs/issues/187
+      hash($x)
+    else:
+      hashData(cast[pointer](unsafeAddr x), type(x).sizeof)
+
+proc hash*[T: SomeNumber | Ordinal | char](x: T): Hash {.inline.} =
+  ## Efficient hashing of numbers, ordinals (eg enum), char.
+  when T.sizeof >= 4:
```
Comment (Clyybber) [inline lib/pure/hashes.nim:130]: I don't entirely understand why this is even needed. Hash is an int, so casting to int should just do it no?

File: lib/pure/collections/sets.nim [hunk 4]
```diff
@@ -1020,7 +1020,7 @@ when isMainModule and not defined(release):
 
     block toSeqAndString:
       var a = toHashSet([2, 7, 5])
-      var b = initHashSet[int]()
+      var b = initHashSet[int](rightSize(a.len))
```
Comment (Clyybber) [inline lib/pure/collections/sets.nim:1022]: Why is this needed?

Comment (timotheecour) [reply lib/pure/collections/sets.nim:1022]: so the test passes, the ordering of elements within the HashSet is implementation defined, eg before this PR:
```nim
import sets
echo toHashSet([1, 2, 3])
echo toHashSet([1, 2, 3, 4])
```
prints:
```
{3, 1, 2}
{3, 2, 1, 4}
```

File: lib/pure/hashes.nim (hunk 3 above)
Comment (timotheecour) [reply lib/pure/hashes.nim:130]: no: with your proposal `proc hash*[T: SomeNumber | Ordinal | char](x: T): Hash {.inline.} = cast[Hash](x)` you get bad results:
* hash(0.0) != hash(-0.0)
* hash(x) is very non random
* hash(x) is doesn't match between RT and CT
```
     0.3 RT => 4599075939470750515
     0.3 CT => 4599075939470750515
     0.0 RT => 0
     0.0 CT => 0
    -0.0 RT => -9223372036854775808
    -0.0 CT => 0
   "foo" RT => 4138058784
   "foo" CT => 4138058784
     123 RT => 123
     123 CT => 123
     124 RT => 124
     124 CT => 124
     125 RT => 125
```

with the algo I have you get:
```
     0.3 RT => -7272369355435918195
     0.3 CT => -7272369355435918195
     0.0 RT => 0
     0.0 CT => 0
    -0.0 RT => 0
    -0.0 CT => 0
   "foo" RT => 4138058784
   "foo" CT => 4138058784
     123 RT => -508526132817443961
     123 CT => -508526132817443961
     124 RT => -1530243182942418667
     124 CT => -1530243182942418667
     125 RT => -6335688243927128111
     125 CT => -6335688243927128111
```

File: lib/pure/collections/sets.nim (hunk 4 above)
Comment (Clyybber) [reply lib/pure/collections/sets.nim:1022]: But the initial size is optional.

File: lib/pure/hashes.nim (hunk 3 above)
Comment (Clyybber) [reply lib/pure/hashes.nim:130]: > hash(x) is very non random

It doesn't need to be random when its perfect.
> hash(0.0) != hash(-0.0)

As expected? Floats work this way. Not the hashsets problem IMO
Why should we construct artificial collisions???
> hash(x) is doesn't match between RT and CT

That can be fixed

File: lib/pure/collections/sets.nim (hunk 4 above)
Comment (timotheecour) [reply lib/pure/collections/sets.nim:1022]: sure but I'm not sure what you're suggesting concretely ; that was the simplest fix here; note that, without adding `rightSize(a.len)` the test is not good anyway, because ordering is implementation defined; so it'll break every time implementation changes; a better test would be to have `toStringSorted(a: HashSet): string` instead of `$a` which would work regardless of implementation

Comment (Clyybber) [reply lib/pure/collections/sets.nim:1022]:
```suggestion
      var b = initHashSet[int]()
```
Thats what I am suggesting : )
I was only referring to this particular line

File: lib/pure/hashes.nim (hunk 3 above)
Comment (timotheecour) [reply lib/pure/hashes.nim:130]: > It doesn't need to be random when its perfect.

this "perfect" identity hash of yours is the exact reason you get slow insertions in https://github.com/nim-lang/Nim/issues/13393; this can result in slowdowns of 100x-1000x, see tests in top post and in that issue

>> hash(0.0) != hash(-0.0)

> As expected? Floats work this way. Not the hashsets problem IMO

sorry, no. I don't know any language where `x == y` is true but hash(x) != hash(y). This makes no mathematical/logical/practical sense.


> > hash(x) is doesn't match between RT and CT

> That can be fixed

indeed; this PR doesn't have this problem

File: lib/pure/collections/sets.nim (hunk 4 above)
Comment (Clyybber) [reply lib/pure/collections/sets.nim:1022]: Ah, I see what you mean now. But the fact that initializing with a initialSize changes the order this way is also implementaion defined.

File: lib/pure/hashes.nim (hunk 3 above)
Comment (Clyybber) [reply lib/pure/hashes.nim:130]: > this "perfect" identity hash of yours is the exact reason you get slow insertions in #13393; this can result in slowdowns of 100x-1000x, see tests in top post and in that issue

No you are wrong: https://github.com/nim-lang/Nim/pull/13410/files#diff-466aee2707291e60fe359bda6b675b86L118
> sorry, no. I don't know any language where x == y is true but hash(x) != hash(y). This makes no mathematical/logical/practical sense.

Oh, I wasn't aware that `0.0 == -0.0` is true.
Ok so we do it your way for floats.

Comment (timotheecour) [reply lib/pure/hashes.nim:130]: > > this "perfect" identity hash of yours is the exact reason you get slow insertions in #13393; this can result in slowdowns of 100x-1000x, see tests in top post and in that issue

> No you are wrong: https://github.com/nim-lang/Nim/pull/13410/files#diff-466aee2707291e60fe359bda6b675b86L118

not sure where I'm wrong; I was showing what happened if I followed (my understanding of) your suggestion `I don't entirely understand why this is even needed. Hash is an int, so casting to int should just do it no?`, so I showed results above for `proc hash*[T: SomeNumber | Ordinal | char](x: T): Hash {.inline.} = cast[Hash](x)`

* this results in predictable, non-random values that cause massive slowdowns during insertions, see benchmarks in https://github.com/nim-lang/Nim/issues/13393 (run it to see for yourself!) and I've already explained why here https://github.com/nim-lang/Nim/pull/11767#issuecomment-512611465.

* using a prime as was done in https://github.com/nim-lang/Nim/pull/13410/files#diff-466aee2707291e60fe359bda6b675b86L118 via `cast[Hash](cast[uint](x) * prime)` didn't properly solve this problem, as evidenced by that benchmark.

* Using this PR, you get a 100x-1000x speedup on this benchmark (which I didn't write btw).

File: lib/pure/collections/sets.nim (hunk 4 above)
Comment (timotheecour) [reply lib/pure/collections/sets.nim:1022]: it's implementation defined but the way i wrote it should work regardless of hash implementation, based on how `toHashSet` is implemented; in any case my main point is that a better test (in general, for any algo testing HashSet, Tables + other structures based on hashes) would involve defining a `itemsSorted(collection)`; left for future PRs

**Comment by timotheecour** [comment]
PTAL:
* added performance regression test that takes into account different input distributions, and on different targets (c, cpp, vm, js)
* improved handling of float and simplified code a bit
* introduced hashUInt64, hashUInt32 which is justified by thashes_perf.nim

File: lib/pure/hashes.nim [hunk 5]
```diff
@@ -86,6 +89,75 @@ proc hashData*(data: pointer, size: int): Hash =
...
+
+proc hashUInt32*(x: uint32): Hash {.inline.} =
+  ## for internal use; user code should prefer `hash` overloads
+  # calling `hashUInt64(x)` would perform 1.736 worse, see thashes_perf toInt32
+  when nimvm: # in vmops
+    doAssert false
+  else:
+    # inspired from https://gist.github.com/badboy/6267743
+    var x = x xor ((x shr 20) xor (x shr 12))
+    # debugEcho ("hashUInt32", x)
```
Comment (narimiran) [inline lib/pure/hashes.nim]: Debugging leftover.

Comment (timotheecour) [reply lib/pure/hashes.nim]: fixed

File: testament/important_packages.nim [hunk 6]
```diff
@@ -36,7 +42,8 @@ pkg "gnuplot"
 pkg "hts", "nim c -o:htss src/hts.nim"
 pkg "illwill", "nimble examples"
 pkg "inim"
-pkg "itertools", "nim doc src/itertools.nim"
+pkg "itertools", "git checkout pr_fix_for_new_hashes2 && nim doc src/itertools.nim".toCommand, url = "https://github.com/timotheecour/itertools"
```
Comment (narimiran) [inline testament/important_packages.nim]: no god no

File: tests/vm/tcompiletimetable.nim [hunk 7]
```diff
@@ -51,8 +51,36 @@ addStuff("Hi"): echo "Hi"
...
 # ensure .compileTime vars can be used at runtime:
-import macros
-
 var xzzzz {.compileTime.}: array[3, string] = ["a", "b", "c"]
 
 for i in 0..high(xzzzz): echo xzzzz[i]
+
+
+import std/hashes
+block:
```
Comment (Araq) [inline tests/vm/tcompiletimetable.nim]: Pointless blocks.

Comment (timotheecour) [reply tests/vm/tcompiletimetable.nim]: fixed

File: testament/important_packages.nim (hunk 6 above)
Comment (timotheecour) [reply testament/important_packages.nim]: I've now removed hotfix now that you've merged it; it's the only way i know of to fix CI (or test a nim PR) before a hotfix is merged upstream

**Comment by Araq** [comment]
Yet there will be other distributions out there that are now much worse than before. The real solution is to add some logic to tables.nim to compensate for the case of very many hash collisions. That's what Python does and Python keeps its hash function for integers really simple.

File: lib/pure/hashes.nim (hunk 3 above)
Comment (Clyybber) [reply lib/pure/hashes.nim:130]: Sorry, I'm stupid :D

**Comment by timotheecour** [comment]
> Yet there will be other distributions out there that are now much worse than before

seems like a purely theoretical argument; while you can in theory find a (contrived, adversarial) distribution that would be made worse by a given hash function, in practice this shouldn't happen, at least much less often than with current hash (but please prove me wrong with a realistic example that causes lots of collisions with new scheme)

Good hash properties are commonly agreed upon, see for eg https://www.sparknotes.com/cs/searching/hashtables/section2/ and https://gist.github.com/badboy/6267743. Of particular importance is avalanche property:
> Avalanche means that a single bit of difference in the input will make about 1/2 of the output bits be different

which the previous hashing scheme we had (`bottom_n_bits(x * 11)`) did not have. It's very easy to find random input distributions that would cause many (or even only) collisions: for example, inputs where only high order bits differ (eg `proc toHighOrderBits(a: int): uint64 = cast[uint64](a) shl 32`), see perf test I've included in this PR.

Another one was small floats, which before this PR would all collapse. These cases do occur in practice, which prompted #13393 and #11764 in the first place. 

> The real solution is to add some logic to tables.nim to compensate for the case of very many hash collisions. That's what Python does and Python keeps its hash function for integers really simple.

we can definitely improve logic for handling collisions, that's not mutually exclusive. But at least with this PR we can avoid those 100x-1000X catastrophic slowdowns from that bug report.

**Comment by Araq** [comment]
>  while you can in theory find a (contrived, adversarial) distribution

Well you said the relevant word here, **adversarial**. This is a problem with hash tables in a stdlib, you don't know what it will be used for but inevitably it will end up for critical software and we need to prevent hash collision attacks. Just coming up with a "better hash" function isn't gonna cut it.

> we can definitely improve logic for handling collisions, that's not mutually exclusive

Yet Python does not need these complex hash functions for integers because its other mechanisms work out. 

Look, I agree that you made collisions less likely for many realworld distributions. But you also made the hashing of integers slower and the bad security is hardly improved.

**Comment by timotheecour** [comment]
> Look, I agree that you made collisions less likely for many realworld distributions. But you also made the hashing of integers slower.

actually that's incorrect, only slower for 64 bit integers:
* for x with sizeof(T)=8 (eg int64), the hashing computation is now 2.2X slower; in practice I doubt the hash computation would be a bottleneck in a real application, where something needs to be done with those keys anyways; the 100x-1000x however is very likely to make a difference.
* for x with sizeof(T)=4 (eg int32), the hash computation is now **1.24X faster** (you can verify by adapting thashes_perf.nim); I obtained this by carefully distinguishing between hashUInt64 vs hashUInt32. Thus, that case should be a no-brainer.
* for x with sizeof(T)<4, no difference

so for performance minded applications, either you have < 4 billions keys and you can benefit from faster hashing via int32/uint32 (and much less chances of catastrophic collisions) [simplifying a bit the argument but you get the idea], or you have > 4 billion keys and need int64, and in that case yes, pay a price for slower key computations but other factors like memory issues (or collisions) would likely be much more important

> Well you said the relevant word here, adversarial.

plenty of applications don't have to deal with adversarial input. And even for the ones that do, there are options to deal with those, eg adding a runtime random seed in hash function, as well as better strategy for dealing with collisions in tables implementation. But reducing chances of collisions in the 1st place should improve performance regardless.

File: testament/important_packages.nim (hunk 6 above)
Comment (narimiran) [reply testament/important_packages.nim]: > I've now removed hotfix now that you've merged i

Can't you in future just wait a bit, instead of flooding our CIs?

**Comment by Araq** [comment]
Your alternative PR was merged, closing.
